### ISSUE TITLE
[tune] Trials stopping early with "loss: 0" 

### ISSUE THREAD

**Issue by trias702** [ray-project/ray#4606]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Ray installed from (source or binary)**: binary
- **Ray version**: 0.6.5
- **Python version**: 3.6.8
- **Exact command to reproduce**:

### Describe the problem
I am trying to implement a basic gridsearch using tune.run on a custom Trainable class I've built. However, every time I run the script, all of my trials are terminating after just one iter:

```
- TrainQTIMDB_0_dropout=0.1,gradient_clip=1.0,learning_rate=0.001,num_layers=1:        TERMINATED, [4 CPUs, 1 GPUs], [pid=2016], 561 s, 1 iter, 6.53 loss
- TrainQTIMDB_1_dropout=0.2,gradient_clip=1.0,learning_rate=0.001,num_layers=1:        TERMINATED, [4 CPUs, 1 GPUs], [pid=2004], 560 s, 1 iter, 6.49 loss
- TrainQTIMDB_2_dropout=0.1,gradient_clip=5.0,learning_rate=0.001,num_layers=1:        TERMINATED, [4 CPUs, 1 GPUs], [pid=1999], 561 s, 1 iter, 6.51 loss

```

I want each trial to train for 3 epochs (3 iters) before it terminates, and the loss reported should be the loss from the final epoch.

I've implemented my Trainable class with the following methods:

```
class TrainQTIMDB(Trainable):
    def _setup(self, config):
         <do some custom setup logic here>

    # this function executes an epoch, should be called 3 times
    def _train_iteration(self, ep):
        self.model.train()

        n_iters = len(self.train_dataloader)

        epoch_loss = 0
        for step, batch in enumerate(self.train_dataloader):
            inputs, batch_lens = batch
            inputs = inputs.to(self.device)
            
            loss = self.model(inputs, batch_lens)
            loss.backward()
            epoch_loss += loss.item()
            ...
        
         return {"mean_loss": epoch_loss}

    def _train(self):
        for ep in range(1, self.args.epochs+1):
            return self._train_iteration(ep)
```

And this is my tune.run code:

```
    tune.run(
        TrainQTIMDB,
        scheduler=None,
        **{
            "name": "TrainQTIMDB_Test",
            "stop": {
                "mean_loss": 0,
                "training_iteration": 3,
            },
            "resources_per_trial": {
                "cpu": 4,
                "gpu": 1,
            },
            "num_samples": 1,
            "checkpoint_at_end": False,
            "config": {
                "args": args,
                "train_data_id": train_data_id,
                "learning_rate": tune.grid_search([0.001, 0.003]),
                "dropout": tune.grid_search([0.1, 0.2]),
                "gradient_clip": tune.grid_search([1., 5.]),
                "num_layers": tune.grid_search([1, 2]),
                #"bidirectional": tune.grid_search([False, True]),
            }
        })
```

I understand that in Trainable, the main entry point function is "_train(self)" and from what I can gather, this function should be called X number of times, where X is the amount of iterations you want each trial to do. I want to do 3 iterations (3 epochs) per trial, but cannot for the life of me seem to make this happen. Hence I manually tried to hack the for epoch loop inside _train() but even that doesn't seem to work.

Can anyone please help?

**Comment by richardliaw** [comment]
Oh; all the stopping criteria assumes monotonicity (this is an issue #4120). Maybe consider setting `neg_mean_loss: 0` instead?

**Comment by trias702** [comment]
Thank you kindly for the link, I understand the issue now.

Please mark this issue as closed.
